### DISCUSSION TITLE
OTA remote - is it available?

### DISCUSSION THREAD

**Discussion by homonto** [HomeSpan/HomeSpan#429] [Ideas]
I am FASCINATED with your library ;-) and your support. Thank you.
I read the OTA section but I cannot find the way of performing remote OTA i.e. the device checks the running sketch version and compares to the one located somewhere on the internet. Or the same but instead of sketch checking it periodically, it could be performed on user action (I am not sure if HomeKit supports "upgrade firmware" button in the app or so).
You know: you install the device in your grandma's house and then you want to upgrade the sketch remotely - without seeing grandma again ;-)

So question: is such functionality not available or am I not reading the documentation properly?

**Comment by HomeSpan** [comment]
Yes, OTA is built into HomeSpan.  Once enabled with within your sketch, you can use the Arduino IDE to upload code directly to the device without plugging it into the serial port.  However, you do have to be on the same WiFi network to have access the device.

To upload from the Arduino IDE, navigate to the *Tools->Port* menu and select your device from the Network Ports section instead of the Serial Ports section.  All HomeSpan devices that have OTA enabled should show up here if they are connected to your WiFi network.  Note that you can use OTA to upload a sketch even if the device is plugged into the serial port, but the Arduino IDE will automatically close the Serial port once you select a Network port.  This is a limitation of the Arduino IDE.

You cannot remotely upload software to another device in grandma's house, unless you connect to her WiFi network.  HomeKit itself does not support software updates through the Home App - it's been a long-standing request and I hope Apple will add something one day.

**Reply by homonto** [reply]
thank you, the first 2 things I am aware of and it works but I was mainly concerned about remote/remote/remote OTA ;-)
and I think I know how to: 
extra button on the device and i.e. 30s push then device executes extra code i.e. download FW from remote internet site
similar approach I have for battery/sleepers/sensors: extra button to wake it upon and then the same as above: different piece of code and download FW from remote site

what do you think? ;-)

**Comment by HomeSpan** [comment]
That could work but you'd need to add specific code to your sketch to download new software from a remote internet site.  I believe the ESP32 IDF has such functionality but I did not integrate anything like this into HomeSpan.

**Reply by homonto** [reply]
such code I have already for my sensors and using button on Home Assistant the guys are connecting remotely, checking if new FW is there and then they perform upgrade
for sleepers as well but extra "push from grandma" is needed so that esp knows that "fw upgrade button" woke it up ;-)

hour by hour I am more fascinated with your library ;-)
<img width="331" alt="Screenshot 2022-11-13 at 15 47 06" src="https://user-images.githubusercontent.com/46562447/201530730-a0da0826-2da1-4d26-aca6-20823115f545.png">
<img width="256" alt="Screenshot 2022-11-13 at 15 47 14" src="https://user-images.githubusercontent.com/46562447/201530736-52384d34-a1fc-461d-a07b-cac8327e832a.png">

**Reply by homonto** [reply]
I ended up "reusing" the "ID" button on Eve to initiate the OTA (for the non-sleeper i.e. Bridge).
It works fine - in the "boolean update()" method I execute firmware upgrade from remote site.
However, Eve shows "ID" button that initiates identify method but I am not able to find such button in HomeKit ;-)

I am wondering, if for HomeKit or Eve we can add any "buttons" or so that would be called "fw upgrade"? or the workaround with reusing "ID" (so: identify) is the only option. See on Eve:
![IMG_0018](https://user-images.githubusercontent.com/46562447/204122791-e31a5099-7da4-4bf7-a8f8-a89e656b5775.jpeg)

**Reply by HomeSpan** [reply]
You can try creating a custom characteristic - see the API for details and an example.  Eve renders custom characteristics in a generic fashion, so you might be able to create your own button that you can use to trigger OTA.
